File: src/main/java/org/jukeboxmc/block/behavior/BlockWater.java
package org.jukeboxmc.block.behavior;

import com.nukkitx.nbt.NbtMap;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.block.BlockType;
import org.jukeboxmc.util.Identifier;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockWater extends BlockLiquid {

    public BlockWater( Identifier identifier ) {
        super( identifier );
    }

    public BlockWater( Identifier identifier, NbtMap blockStates ) {
        super( identifier, blockStates );
    }

    @Override
    public boolean canBeReplaced( Block block ) {
        return true;
    }

    @Override
    public int getTickRate() {
        return 5;
    }

    @Override
    public boolean usesWaterLogging() {
        return true;
    }

    @Override
    public boolean canPassThrough() {
        return true;
    }

    @Override
    public BlockLiquid getBlock( int liquidDepth ) {
        BlockWater blockWater = Block.create( BlockType.WATER );
        blockWater.setLiquidDepth( liquidDepth );
        return blockWater;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockWater.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.Liquid
import org.jukeboxmc.api.block.Water
import org.jukeboxmc.api.block.data.BlockFace
import org.jukeboxmc.api.math.Vector
import org.jukeboxmc.server.block.JukeboxBlock
import org.jukeboxmc.server.item.JukeboxItem
import org.jukeboxmc.server.player.JukeboxPlayer
import org.jukeboxmc.server.world.JukeboxWorld

open class BlockWater(identifier: Identifier, blockStates: NbtMap?) : BlockLiquid(identifier, blockStates), Water {

    override fun placeBlock(
        player: JukeboxPlayer,
        world: JukeboxWorld,
        blockPosition: Vector,
        placePosition: Vector,
        clickedPosition: Vector,
        itemInHand: JukeboxItem,
        blockFace: BlockFace
    ): Boolean {
        world.setBlock(placePosition.getBlockX(), placePosition.getBlockY(), placePosition.getBlockZ(), 0, placePosition.getDimension(), this, true)
        return true
    }

    override fun getLiquidWithNewDepth(depth: Int): Liquid {
        return (blockWater.clone() as BlockWater).setLiquidDepth(depth)
    }

    override fun canCollideCheck(block: JukeboxBlock, value: Boolean): Boolean {
        if (block is BlockWater) {
            return value && block.getLiquidDepth() == 0
        }
        return false
    }

    override fun tickRate(): Long {
        return 5
    }
}